website/src/content/development/forms.mdx

```diff
@@ -0,0 +1,540 @@
+
+```js highlightLines="6-7"
+<TextField
+  name="key"
+  title="Key"
+  isRequired={true}
+  value={formik.values.key}
+  errors={formik.errors.key}
+  touched={formik.touched.key}
+  onChange={formik.handleChange}
+  onBlur={formik.handleBlur}
+/>
+```
+
+<Info>
+
+Error messages are only shown when the `touched` value for the specific field is set to `true`. This happens whenever the user stops interacting with a field (loses `focus`). The point here is that as long as the user is interacting with a field for the first time there is no need to show a validation.
+
+</Info>
+
+By default field components have built-in error messages for the `missing` error key. Any other error message can be mapped and rendered using the `renderError` function.
+
+```js highlightLines="7-14"
+<TextField
+  name="key"
+  title="Key"
+  isRequired={true}
+  value={formik.values.key}
+  errors={formik.errors.key}
+  renderError={(errorKey) => {
+    switch (errorKey) {
+      case 'invalid':
+        return 'The value is invalid.';
+      default:
+        return null;
+    }
+  }}
+  touched={formik.touched.key}
+  onChange={formik.handleChange}
+  onBlur={formik.handleBlur}
+/>
+```
+
+# Form data conversion
+
+Implementing forms is almost always related to managing data as a bidirectional data flow.
+
+image
+
+To facilitate converting data to and from a form, we recommend to define some conversion functions.
+
+* `docToFormValues`: converts data, for example fetched from an API, to the form specific format.
+* `formValuesToDoc`: converts form data back to the original data format, for example to be used in an API.
+
+The `docToFormValues` is what you would usually use for initializing the form.
+
+```jsx highlightLines="2,6"
+import { useFormik } from 'formik';
+import { docToFormValues } from './conversions';
+
+const ChannelsForm = () => {
+  const formik = useFormik({
+    initialValues: docToFormValues(props.data),
+    // ...
+  });
+  // ...
+}
+```
+
+In our Channels example, it can look something like this:
+
+```js title="conversions.js"
+export const docToFormValues = (doc) => ({
+  key: doc?.key ?? '',
+})
+```
+
+The `formValuesToDoc` on the other end is what you would usually use when submitting the form.
+
+```jsx highlightLines="2,7"
+import { useFormik } from 'formik';
+import { formValuesToDoc } from './conversions';
+
+const ChannelsForm = () => {
+  const formik = useFormik({
+    onSubmit: async (formValues) => {
+      const updateData = formValuesToDoc(formValues);
+    },
+    // ...
+  });
+  // ...
+}
+```
+
+In our Channels example, it can look something like this:
+
+```js title="conversions.js"
+export const formValuesToDoc = (formValues) => ({
+  key: formValues.key,
+})
+```
+
+<Info>
+
+Most of the time you would have a simple 1:1 mapping. However, we still recommend to use these conversion functions as a best practice and to help decoupling the data transformation logic from the form component.
```

Comment:
Maybe add that it also helps not sending uninteresting fields to a form down. People otherwise might send a while product of the API into a form which intends to work on two fields only.

Reply:
Good point, thanks!